addons/binding/org.openhab.binding.enigma2/src/main/java/org/openhab/binding/enigma2/internal/Enigma2Util.java

```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    /**
+     * Executes an URL and returns to answer
+     */
+    public static String executeUrl(String url) throws IOException {
+        return HttpUtil.executeUrl("GET", url, 5000);
+    }
+
+    /**
+     * Scans "http://enigma2/web/getallservices" and generates map
+     *
+     * @throws IOException
+     * @throws ParserConfigurationException
+     * @throws SAXException
+     */
+    public static Enigma2ServiceContainer generateServiceMaps(String deviceURL)
+            throws IOException, ParserConfigurationException, SAXException {
+        Enigma2ServiceContainer serviceContainer = new Enigma2ServiceContainer();
+
+        File inputFile = new File("services.xml");
+        BufferedWriter writer = new BufferedWriter(new FileWriter(inputFile));
+
+        String content = executeUrl(deviceURL + SUFFIX_ALL_SERVICES);
+        writer.write(content);
+        writer.close();
+
+        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
+        DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
+        Document doc = dBuilder.parse(inputFile);
+        doc.getDocumentElement().normalize();
+        NodeList listOfBouquets = doc.getElementsByTagName("e2bouquet");
+        for (int bouquetIndex = 0; bouquetIndex < listOfBouquets.getLength(); bouquetIndex++) {
+            NodeList listOfServices = doc.getElementsByTagName("e2servicelist");
+            for (int serviceIndex = 0; serviceIndex < listOfServices.getLength(); serviceIndex++) {
+                NodeList serviceList = doc.getElementsByTagName("e2service");
+                for (int i = 0; i < serviceList.getLength(); i++) {
+                    Node service = serviceList.item(i);
+                    if (service.getNodeType() == Node.ELEMENT_NODE) {
+                        Element eElement = (Element) service;
+                        String e2servicereference = eElement.getElementsByTagName("e2servicereference").item(0)
+                                .getTextContent();
+                        String e2servicename = eElement.getElementsByTagName("e2servicename").item(0).getTextContent();
+
+                        serviceContainer.add(e2servicename, e2servicereference);
+                    }
+                }
+            }
+        }
+        inputFile.delete();
+        return serviceContainer;
+    }
+
+    public static String createUserPasswordHostnamePrefix(String hostName, String userName, String password) {
+        String returnString;
+        if ((userName == null) || (userName.length() == 0)) {
+            returnString = new StringBuffer("http://" + hostName).toString();
+        } else {
+            returnString = new StringBuffer("http://" + userName).append(":").append(password).append("@")
+                    .append(hostName).toString();
+        }
+        return returnString;
+    }
+
+    public static String cleanString(String string) {
+        StringBuffer buffer = new StringBuffer();
```

Comment:
`StringBuilder` is faster than `StringBuffer` because it's not synchronized. So you should always use `StringBuilder`
 
And the thread safety which is delivered by `StringBuffer` is usually not enough because different append's can still be interleaved which still ruin your `StringBuffer`.

Just memorize to use the one which is the furthest in the Alphabet.